### PR TITLE
device: show address on device display

### PR DESCRIPTION
**Pull request by ph4r05**
- Trezor: support for device address display (subaddress, integrated address)
- Wallet::API support added
- Simplewallet:
  - `address device [<index>]`
  - `address new <label>`  // shows the address also on the device
  - `integrated_address [device] <payment_id|address>`  // new optional "device" arg to display also on the device

Relates to:
- https://github.com/monero-project/monero-gui/issues/2184
- https://github.com/trezor/trezor-firmware/pull/201
- https://www.reddit.com/r/Monero/comments/c1mgzm/how_to_verify_subaddress_on_trezorledger_device/

### CODE REVIEW

File: src/device_trezor/device_trezor.cpp [hunk 1]
```diff
@@ -209,15 +214,29 @@ namespace trezor {
...
 
       auto req = std::make_shared<messages::monero::MoneroGetAddress>();
       this->set_msg_addr<messages::monero::MoneroGetAddress>(req.get(), path, network_type);
+      req->set_show_display(show_address);
+      if (subaddress){
+        req->set_account(subaddress->major);
+        req->set_minor(subaddress->minor);
+      }
+
+      // TODO: add once PR in trezor-firmware is merged
```
Comment (ph4r05) [inline src/device_trezor/device_trezor.cpp]: We need to wait https://github.com/trezor/trezor-firmware/pull/201 to be merged

**Comment by ph4r05** [comment]
btw @moneromooo-monero is there pls an easy way to get the current hardfork version being used? 

I've checked existing methods. Usually, a caller provides HF version. Thanks for an advice.

**Comment by moneromooo-monero** [comment]
From the wallet ? Not currently. I have a commit somewhere that exposes this though. Gimme a min...

**Comment by moneromooo-monero** [comment]
https://github.com/moneromooo-monero/bitmonero/commit/694420b0370c7df044feb2295034ddab1b22cfd5

**Comment by moneromooo-monero** [comment]
If it's useful generally, I can PR it and add a "get current fork" function in the wallet.

**Comment by ph4r05** [comment]
> If it's useful generally, I can PR it and add a "get current fork" function in the wallet.

That would be wonderful, thanks!

Our use case is to send current hardfork version the transaction is being made for, so if the Monero codebase is updated without Trezor client support being compatible, the Trezor should deny signing the transaction.

**Comment by ph4r05** [comment]
But this PR can be merged before that, independently :)

**Comment by moneromooo-monero** [comment]
The wallet has a use use_fork_rules function. That's probably what you want.

**Comment by ph4r05** [comment]
I’ve checked the use_fork_rules. It takes hf version ans returns boolean. I would like to have a function returning the hf version.

But I could do it naively - go from the latest hf and test with use_fork_rules. Is linear in calls byt it works.

**Comment by moneromooo-monero** [comment]
Why would you want to check for obsolete stuff ?

**Comment by ph4r05** [comment]
> Why would you want to check for obsolete stuff ?

The idea is to get the most recent HF being used without the need to modify Trezor client code. So if HF increases and we somehow don't update the client code Trezor device will refuse to sign.

**Comment by moneromooo-monero** [comment]
So it'd be the firmware itself checking ?

**Comment by ph4r05** [comment]
> So it'd be the firmware itself checking ?

Correct. We would like to have a fail-safe check in the Trezor firmware for the situation the Monero code switches the hardfork before us knowing and adapting both firmware and Trezor code in the Monero codebase.

**Comment by moneromooo-monero** [comment]
OK, then the patch I posted earlier does not do what you want. What you want is calling hard_fork_info on the daemon with version == 0, and read version from the result.

**Comment by ph4r05** [comment]
Thanks for the suggestion. I've force-pushed the update.

File: src/device/device_cold.hpp [hunk 2]
```diff
@@ -47,6 +47,7 @@ namespace hw {
     std::vector<cryptonote::address_parse_info> tx_recipients;  // as entered by user
     boost::optional<int> bp_version;  // BP version to use
     boost::optional<unsigned> client_version;  // Signing client version to use (testing)
+    boost::optional<unsigned> hard_fork;  // hard fork being used for the transaction
```
Comment (moneromooo-monero) [inline src/device/device_cold.hpp]: Existing code uses uint8_t for this. Might as well use the same type.

File: src/device/device.hpp [hunk 3]
```diff
@@ -236,6 +236,7 @@ namespace hw {
         virtual bool  compute_key_image(const cryptonote::account_keys& ack, const crypto::public_key& out_key, const crypto::key_derivation& recv_derivation, size_t real_output_index, const cryptonote::subaddress_index& received_index, cryptonote::keypair& in_ephemeral, crypto::key_image& ki) { return false; }
         virtual void  computing_key_images(bool started) {};
         virtual void  set_network_type(cryptonote::network_type network_type) { }
+        virtual void  display_address(uint32_t account_index, uint32_t address_index, const boost::optional<crypto::hash8> &payment_id) {}
```
Comment (moneromooo-monero) [inline src/device/device.hpp]: Why not pass the subaddress_index instead of two ints ?

**Comment by ph4r05** [comment]
Done :)

**Comment by ph4r05** [comment]
I was wondering, does the PR look OK? 🙏

File: src/simplewallet/simplewallet.cpp [hunk 4]
```diff
@@ -8585,6 +8587,27 @@ bool simple_wallet::print_address(const std::vector<std::string> &args/* = std::
...
+    if (local_args.size() > 0)
+    {
+      if (!epee::string_tools::get_xtype_from_string(index, local_args[0]))
+      {
+        fail_msg_writer() << tr("failed to parse index: ") << local_args[0];
+        return true;
+      }
+      if (index >= m_wallet->get_num_subaddresses(m_current_subaddress_account))
+      {
+        fail_msg_writer() << tr("<index> is out of bound");
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: bounds

Comment (ph4r05) [reply src/simplewallet/simplewallet.cpp]: fixed! :)

**Comment by ph4r05** [comment]
Thanks!

**Comment by ph4r05** [comment]
@moneromooo-monero ah sry, I figured out there was a minor bug in the wallet::API. Its fixed now

**Comment by ph4r05** [comment]
Ah sorry again, I was integrating the feature to monero-gui and realized I need empty payment id.

**Comment by ph4r05** [comment]
@fluffypony do you pls plan to merge this month? :)

**Comment by moneromooo-monero** [comment]
After he's done the 0.14.1.1 release (and merged the relevant patches for it) please.

**Comment by ph4r05** [comment]
Aaah thanks for info!

**Pull request merged**